src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,566 @@
+ * identical result.
+ *
+ * <p>Implementations of Gatherer must not capture, retain, or expose to
+ * other threads, the references to the state instance, or the downstream
+ * {@link Downstream} for longer than the invocation duration of the method
+ * which they are passed to.
+ *
+ * <p>Performing a gathering operation with a {@code Gatherer} should produce a
+ * result equivalent to:
+ *
+ * <pre>{@code
+ *     Gatherer.Downstream<? super R> downstream = ...;
+ *     A state = gatherer.initializer().get();
+ *     for (T t : data) {
+ *         gatherer.integrator().integrate(state, t, downstream);
+ *     }
+ *     gatherer.finisher().accept(state, downstream);
+ * }</pre>
+ *
+ * <p>However, the library is free to partition the input, perform the
+ * integrations on the partitions, and then use the combiner function to
+ * combine the partial results to achieve a gathering operation.  (Depending
+ * on the specific gathering operation, this may perform better or worse,
+ * depending on the relative cost of the integrator and combiner functions.)
+ *
+ * <p>In addition to the predefined implementations in {@link Gatherers}, the
+ * static factory methods {@code of(...)} and {@code ofSequential(...)}
+ * can be used to construct gatherers.  For example, you could create a gatherer
+ * that implements the equivalent of
+ * {@link java.util.stream.Stream#map(java.util.function.Function)} with:
+ *
+ * <pre>{@code
+ * public static <T, R> Gatherer<T, ?, R> map(Function<? super T, ? extends R> mapper) {
+ *     return Gatherer.of(
+ *         (unused, element, downstream) -> // integrator
+ *             downstream.push(mapper.apply(element))
+ *     );
+ * }
+ * }</pre>
+ *
+ * <p>Gatherers are designed to be <em>composed</em>; two or more Gatherers can
+ * be composed into a single Gatherer using the {@link #andThen(Gatherer)}
+ * method.
+ *
+ * <pre>{@code
+ * // using the implementation of `map` as seen above
+ * Gatherer<Integer, ?, Integer> increment = map(i -> i + 1);
+ *
+ * Gatherer<Object, ?, String> toString = map(i -> i.toString());
+ *
+ * Gatherer<Integer, ?, String> incrementThenToString = plusOne.andThen(intToString);
+ * }</pre>
+ *
+ * AS an example, in order to create a gatherer to implement a sequential
+ * Prefix Scan as a Gatherer, it could be done the following way:
+ *
+ * <pre>{@code
+ * public static <T, R> Gatherer<T, ?, R> scan(
+ *     Supplier<R> initial,
+ *     BiFunction<? super R, ? super T, ? extends R> scanner) {
+ *
+ *     class State {
+ *         R current;
+ *         State() {
+ *             current = initial.get();
+ *         }
+ *     }
+ *
+ *     return Gatherer.<T, State, R>ofSequential(
+ *          State::new,
+ *          Gatherer.Integrator.ofGreedy((state, element, downstream) -> {
+ *              state.current = scanner.apply(state.current, element);
+ *              return downstream.push(state.current);
+ *          })
+ *     );
+ * }
```

Comment:
As a matter of style, I find it weird that `current` is assigned in both the constructor and in the lambda.  I'd be inclined to put all assignments in the State class (and expose more methods) or put all assignments in the lambdas (and ditch the ctor).  For an example this trivial, it doens't make so much differnce, but it gives a gentle push towards doing things more consistenly.

Reply:
making `integrate` a method on State in this case is a bit of extra noise, which I wanted to avoid for something embedded in the Javadoc. I'll reduce some of the noise by assigning `current` in the member declaration instead of in an explicit ctor.